src/hiccup/compiler.clj

```diff
@@ -314,22 +333,24 @@
 (defmethod compile-element ::literal-tag
   [[tag attrs & content]]
   (let [[tag tag-attrs _] (normalize-element-form [tag])
-        attrs-sym         (gensym "attrs")]
+        attrs-sym (gensym "attrs")]
     `(let [~attrs-sym ~attrs]
-       (if (map? ~attrs-sym)
-         ~(if (container-tag? tag content)
-            `(build-string ~(str "<" tag)
-                           (render-attr-map (merge ~tag-attrs ~attrs-sym)) ">"
-                           ~@(compile-seq content)
-                           ~(str "</" tag ">"))
-            `(build-string ~(str "<" tag)
-                           (render-attr-map (merge ~tag-attrs ~attrs-sym))
-                           ~(end-tag)))
-         ~(if (container-tag? tag attrs)
-            `(build-string ~(str "<" tag (render-attr-map tag-attrs) ">")
-                           ~@(compile-seq (cons attrs-sym content))
-                           ~(str "</" tag ">"))
-            (build-string "<" tag (render-attr-map tag-attrs) (end-tag)))))))
+       (build-string
+         (if (map? ~attrs-sym)
+           ~(if (container-tag? tag content)
+              `(build-string ~(str "<" tag)
+                             (render-attr-map (merge ~tag-attrs ~attrs-sym))
+                             ">")
+              `(build-string ~(str "<" tag)
+                             (render-attr-map (merge ~tag-attrs ~attrs-sym))
+                             ~(end-tag)))
+           (build-string ~(str "<" tag (render-attr-map tag-attrs) ">")
+                         ~@(compile-seq [attrs-sym])))
+         ;; content and ending tag (when the above code did not emit an ending tag)
+         (when (or ~(boolean (container-tag? tag content))
+                   (not (map? ~attrs-sym)))
```

Comment:
Could this be written as:

```clojure
~(if (container-tag? tag content)
   `(build-string ~@(compile-seq content) ~(str "</" tag ">"))
   `(when (not (map? ~attrs-sym))
      (build-string ~@(compile-seq content) ~(str "</" tag ">"))))
```

Unless I've misread the code, this should do the same thing, but potentially shave off another form in some circumstances.